docs/2019-05-23-Unification-Modulo-Equations.md

```diff
@@ -0,0 +1,163 @@
+Unification with equations
+==========================
+
+Summary
+-------
+
+Proposal: Write k/kore axioms/theorems allowing us both to solve unification
+in a principled way without changing the backend for each of these cases,
+and to easily generate proofs if/when that will be desirable.
+
+Background
+----------
+
+The most general unifier of `φ` and `ψ` is `⌈φ∧ψ⌉`
+
+There are some unification cases that are easy to solve, e.g. if `C` is a
+constructor (actually, it only needs to be injective), then
+```
+⌈C(φ_1, …, φ_n) ∧ C(ψ_1, …, ψ_n)⌉=⌈φ_1∧ψ_1⌉∧…∧⌈φ_n∧ψ_n⌉
+```
+
+However, if `f` and `g` are random symbols, without any properties, then
+`⌈f(φ_1, …, φ_n) ∧ g(ψ_1, …, ψ_m)⌉` cannot be reduced to a nicer form.
+
+If `f` is, say, an associative binary operation with neutral element (i.e. the
+concatenation for a list), then we have a reasonable intuition about how to
+unify it, but I don’t know of any written text about that.
+
+Problem
+-------
+
+We are trying to extend unification to sorts that do not have constructors, but
+which have something resembling constructors. An example is a list, which has:
+`unit` - the empty list.
+`elem` - takes an element of a different sort and produces a single element list.
+`concat` - concatenates two lists.
+Heating and cooling could be another example.
+
+Proposed Solution
+-----------------
+
+The first attempt to solve this would be to define some sort of
+no-junk/no-confusion axiom
+for that sort (e.g. any list can be built with `unit`, `elem` and `concat`),
+and then expand the two terms in all possible ways using the available
+equations, then match the two terms at the top as long as we stay within
+the same sort, then use unification for the remainders.
+
+However, this has a few problems. First, there is no way of proving that
+this is the most general unifier, i.e. there is no way of showing that we
+truly expanded the term in all possible ways. Second, this would be really
+inefficient.
+
```

Comment:
Another problem is that the (No Confusion) axioms for symbols with underlying axioms might take complex forms (e.g., conder the (No Confusion) axioms for `concat`).